### PR TITLE
Add "# fmt: skip" directive to black

### LINKED ISSUE #1162: # fmt: pass - ignore formatting for single statements
Feature request:
`# fmt: pass` would skip formatting for the single statement that immediately follows.

**Is your feature request related to a problem? Please describe.** 

Not a response to a problem but it might be helpful for the following reasons:

- Easier to use / less typing
- Can't forget to turn it back on
  - Or accidentally insert code you *do* want formatted into an unformatted block
- Encourages ignoring less code
  - One-offs are always isolated to one statement instead of potentially large chunks of code.

**Describe the solution you'd like**

```python
z = "hello " + "world"

# fmt: pass
custom_formatting = [
    0,  1,  2,
    3,  4,  5,
    6,  7,  8,
]

regular_formatting = [
   0,  1,  2,  'this will be formatted'
]
```

The `custom_formatting` list would be ignored, everything else would be formatted.

**Describe alternatives you've considered**

1. Current solution

```python
# fmt: off
ugly_code=('but I want it this way')
even_more=('dont\'t judge me')
# fmt: on
```
2. `#fmt: off` at end of line (#790)
   I believe `# fmt: pass` addresses the same concern while also being a bit more flexible.
    - a single statement could be multi-line
    - a long line would bury an inline comment
    - `# fmt: pass` is more readable & predictable as a separate line
3. Disable formatting for certain expressions only. #798
4. Deal with it

**Additional context**

Naming:

- `# fmt: pass` reflects Python's existing [pass statement](https://docs.python.org/3/reference/simple_stmts.html?highlight=pass#the-pass-statement). Hopefully this hints that you can use it to temporarily appease the formatter for one quick thing, but it's not something you want everywhere.

- Naming could be anything, e.g. "pass", "skip", "ignore"

### PR DESCRIPTION
**Pull request by saroad2**
Fixes #1162

After being requested for quite some time, implemented the ability to skip formatting only one line using `#fmt: skip` 😄 .

### CODE REVIEW

**Review by cooperlees — APPROVED** [review]
Thanks! Code makes sense to me, and the tests look to have good coverage. This looks good to me.

**Comment by ichard26** [comment]
The lack of documentation additions scares me :-) I would like to see a quick explanation on how it works and a few examples. Although honestly the documentation needs to be restructured again since it's so messy right now. I don't know about my availability but I'll probably end up being the guy who will handle this refactor. See #1759.

**Comment by cooperlees** [comment]
Yeah my bad there. Good call @ichard26 - If you make a plan and split up how you'd like things refactored we might be able to get help from others on that so you don't have to do it all!

**Comment by saroad2** [comment]
Hey guys, you want me to fill the documentation gaps, or you're on it?

**Comment by HarrySky** [comment]
Great feature! What is the status of this PR? Friendly ping @ichard26 @cooperlees

**Comment by mrgnw** [comment]
I got this error, looks like the first pass removes `# fmt: skip`.
```sh
error: cannot format blackskip.py: INTERNAL ERROR: Black produced different code on the second pass of the formatter.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful:
```

Log:
```diff
Mode(target_versions=set(), line_length=88, string_normalization=True, experimental_string_processing=False, is_pyi=False)
--- source
+++ first pass
@@ -1,11 +1,8 @@
-z = 'hello ' + "world"
+z = "hello " + "world"
 
-# fmt: skip
 custom_formatting = [
     0,  1,
     2
 ]
 
-regular_formatting = [
-   0,  1,  2,  'this will be formatted'
-]
+regular_formatting = [0, 1, 2, "this will be formatted"]
--- first pass
+++ second pass
@@ -1,8 +1,5 @@
 z = "hello " + "world"
 
-custom_formatting = [
-    0,  1,
-    2
-]
+custom_formatting = [0, 1, 2]
 
 regular_formatting = [0, 1, 2, "this will be formatted"]
```
(Python 3.8.0 / macOS 10.15, `pip install git+git://github.com/saroad2/black.git@add_skip`)

**Comment by saroad2** [comment]
First of all, nice catch @mrgnw ! I fixed the bug and added a unit test to validate this won't happen again.

However, this is a misuse of the `# fmt: skip` directive. You should use this directive at the end of the line you wish to skip. For example:

```python
l = ["I", "will", "not", "be", "formatted",]  # fmt: skip
```

The following **will not work**:

```python
# fmt: skip
l = ["I", "will", "be", "formatted",] 
```

**Comment by mrgnw** [comment]
I see, thank you! I just copy+pasted the example from #1162. Technically I think this is more in line with #790 because it's in-line, but I think this is great and we'd all be happy to have it.

**Comment by saroad2** [comment]
Kindly reminding that this PR is ready to be merged.

**Comment by JelleZijlstra** [comment]
Could you add tests that put `# fmt: skip` in the middle of a complicated statement like a multi-line if condition? That's where we got most of our stability bugs lately, so it would be good to make sure this new feature doesn't introduce more.

I'm supportive of merging otherwise.

**Comment by saroad2** [comment]
Hey @JelleZijlstra , thank you for the response!
Apparently, there was actually a bug there! nice catch!

I fixed the bug and added another unit test to check the new behavior. If you have some other edge cases you think we should address, please let me know.

**Comment by teresaejunior** [comment]
Sorry for the +1, but since this PR is ready, it would come in very handy for me, since I'm in the process of sprinkling my code with debug lines that are to be stripped on code optimization with `python -O`, like:
```py
if __debug__: log.debug(f"{repr(big_dict)[:100]=}") # fmt: skip
```
black reformats it into two lines.

**Comment by saroad2** [comment]
Hey @cooperlees , @ichard26 , @JelleZijlstra, and maybe other collaborators of the Black project.

This PR is almost 4 months old. I know you are busy people, and I know that this a voluntary project, but I would be very disappointed if a new version of Black would be released without it.

So please, if something is missing in this PR, let me know and I'll fix/add/update it.
If not, let's merge this thing!

Again, I must emphasize that I don't mean to harm anyone or to be disrespectful. I just want to see my work being used :)

**Comment by ichard26** [comment]
TBH, I'm a tiny bit hesitant since this _may_ go against of Black's goal of being consistent, but since we have `fmt: off/on` already and while this will probably mean less consistent code since `# fmt: skip` is convenient, the benefits seem worth it. The only actual thing delaying/blocking this is that this should have documentation. I've given up on finishing up the reorganization effort before this PR gets merged since it's 99.9999999% not gonna happen due to an aforementioned lack of time. Aplogies for not getting back to you sooner. For now, I recommend adding the docs [to this section](https://github.com/psf/black/blob/master/README.md#the-black-code-style). I can deal with the messiness myself, eventually.

**Comment by saroad2** [comment]
Hey @ichard26 , thanks for the feedback!

I added the missing documentation piece to the README file.
As for the build failure, I have no idea what causes it, but it doesn't seem like my code.

Let me know if something else is missing!

**Comment by JelleZijlstra** [comment]
Merged. I guess I was hesitant because I'm still worried this will cause new stability bugs. I guess we'll see. Thanks for your contribution to Black!

**Comment by teresaejunior** [comment]
It doesn't work for me. In an empty file with just `if True: pass  # fmt: skip`, a newline gets appended, and if the space is omitted between `#` and `fmt`, it generates invalid code.

Tested on Python 3.9.1, black 20.8b2.dev83+g71bbb67

**Comment by ptsl** [comment]
From the documentation:
> lines that ends with # fmt: skip
Does this mean that `# fmt: skip` cannot be used as a single directive on the line just before the statement (as requested in Issue #1162)?

**Comment by ichard26** [comment]
> Does this mean that # fmt: skip cannot be used as a single directive on the line just before the statement (as requested in Issue #1162)?

I suppose so. /cc @saroad2 was this an intentional choice due to technicalities or something else?

**Comment by saroad2** [comment]
I guess I have linked the wrong issue to this PR.

The idea of this PR was to ignore lines which **ends with**  ` # fmt: skip`. For example, the following line will not be formatted:
```python
l = ["I have",    "too much space"]  # fmt: skip
```

However, the following **will be formatted**:
```python
# fmt: skip
l = ["I have",    "too much space"]
```

A new PR might include the ability that if `# fmt: skip` is added to a blank line, the **next** line should not be formatted. I guess it should be relatively easy to do, and if you wish to, I'll go ahead and make this PR.

Another thing that was mentioned here in the comments is that in order to check which lines to skip, *Black* **compare** the comment to specific values. Therefore, if you miss a space as @teresaejunior said, It will not ignore that line. Another PR should use **regex comparison** instead in order to solve this problem.

**Comment by teresaejunior** [comment]
@saroad2 It could be my specific Python version, but it doesn't work even if the space is included. And instead of ignoring `#fmt` as it should, it produces invalid code. It is always reproducible for me, and I don't have other problems with black with the Python 3.9.1 / black 20.8b2.dev83+g71bbb67 combo.

**Comment by ichard26** [comment]
@teresaejunior the newline handling being completely unaffected issue has [been also reported](https://github.com/psf/black/issues/496) with the old-style `fmt: off/on`.

**Comment by teresaejunior** [comment]
@ichard @saroad2 Ah, OK, it works for other kinds of formatting! (Sorry, I don't know if I should be pinging you.)

**Comment by HarrySky** [comment]
Hi, great to see this feature merged, but it seems that it cannot be used with other comments like "noqa"

This will be skipped (as expected):
```python
log.debug("Long debug message")  # fmt: skip
```

This will be formatted:
```python
log.debug("Long debug message, so we ignore flake8's E501 long line message")  # fmt: skip # noqa: E501
```

Should I create an issue about it? Or is it how `black` is supposed to work?

For example `bandit` and `flake8` detect everything properly if I put both `# nosec # noqa` comments on the same line.

**Comment by JelleZijlstra** [comment]
Please open a new issue. I feel like it should work as you expect, but others may disagree.

**Comment by bersbersbers** [comment]
> Please open a new issue.

For any one (like me) searching: https://github.com/psf/black/issues/2213
